test/extensions/common/aws/metadata_fetcher_test.cc

```diff
@@ -0,0 +1,301 @@
+}
+
+TEST_F(MetadataFetcherTest, TestClusterNotFound) {
+  // Setup without thread local cluster
+  fetcher_ = MetadataFetcher::create(mock_factory_ctx_.cluster_manager_, "cluster_name");
+  Http::RequestMessageImpl message;
+  MockMetadataReceiver receiver;
+
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_, getThreadLocalCluster(_))
+      .WillOnce(Return(nullptr));
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::MissingConfig));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestCancel) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+  Http::MockAsyncClientRequest request(
+      &(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_));
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, &request);
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(request, cancel());
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(testing::_)).Times(0);
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+  // Proper cancel
+  fetcher_->cancel();
+  // Re-entrant cancel
+  fetcher_->cancel();
+}
+
+TEST_F(MetadataFetcherTest, TestSpanPassedDown) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "200", "not_empty");
+  MockMetadataReceiver receiver;
+
+  // Expectations for span
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_,
+              send_(_, _, _))
+      .WillOnce(Invoke(
+          [this](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks&,
+                 const Http::AsyncClient::RequestOptions& options) -> Http::AsyncClient::Request* {
+            EXPECT_TRUE(options.parent_span_ == &this->parent_span_);
+            EXPECT_TRUE(options.child_span_name_ == "AWS Metadata Fetch");
+            return nullptr;
+          }));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestDefaultRetryPolicy) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "200", "not_empty");
+  MockMetadataReceiver receiver;
+
+  EXPECT_CALL(mock_factory_ctx_.cluster_manager_.thread_local_cluster_.async_client_,
+              send_(_, _, _))
+      .WillOnce(Invoke(
+          [](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks&,
+             const Http::AsyncClient::RequestOptions& options) -> Http::AsyncClient::Request* {
+            EXPECT_TRUE(options.retry_policy.has_value());
+            EXPECT_TRUE(options.buffer_body_for_retry);
+            EXPECT_TRUE(options.retry_policy.value().has_num_retries());
+            EXPECT_EQ(PROTOBUF_GET_WRAPPED_REQUIRED(options.retry_policy.value(), num_retries), 3);
```

Comment:
This seems redundant - PROTOBUF_GET_WRAPPED_REQUIRED is just doing the same thing as ASSERT_TRUE would have. I'd suggest switching EXPECT_TRUE for ASSERT_TRUE and then doing `EXPECT_EQ(options.retry_policy.value().num_retries(), 3)`

(Plus other instances of same.)